File: public/data/chapters/chapter_01_arithmetic/logic_advanced.R
# Logic for R Runtime
# Supports add, mul, sin, cos
# No external dependencies (jsonlite) to ensure portability

args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 1) {
  cat("Error: No context provided\n", file = stderr())
  quit(status = 1)
}

json_str <- args[1]

get_val <- function(json, key, type = "numeric") {
  # Looking for "key": value OR "key":value
  # Value can be number, string ("..."), or null
  pattern <- paste0('"', key, '"\\s*:\\s*([^,}]+)')
  match <- regexec(pattern, json)
  result <- regmatches(json, match)[[1]]

  if (length(result) < 2) {
    return(NULL)
  }

  val_str <- result[2]
  # Clean up quotes for strings
  val_str <- gsub('^"|"$', "", val_str)
  val_str <- trimws(val_str)

  if (type == "numeric") {
    if (val_str == "null") {
      return(0)
    } # Handle null as 0
    return(as.numeric(val_str))
  } else {
    return(val_str)
  }
}

# Check for batch mode
is_batch <- grepl('"batch"\\s*:\\s*true', json_str)

if (is_batch) {
  # Parse examples list manually
  # "examples": [ { ... }, { ... } ]

  # Find start of examples array
  ex_match <- regexpr('"examples"\\s*:\\s*\\[', json_str)
  if (ex_match == -1) {
    cat("[]")
    quit(status = 0)
  }

  # Extract content inside [...]
  start_pos <- ex_match + attr(ex_match, "match.length")
  content <- substring(json_str, start_pos)
  # Basic bracket matching or just take until end (assuming simple context)

  # Split by "},{" to separate objects
  # This is fragile but sufficient for controlled test inputs
  items <- strsplit(content, "\\},\\{")[[1]]

  results <- c()

  for (item in items) {
    # Ensure item has braces for regex to work
    if (!grepl("^\\{", item)) item <- paste0("{", item)
    if (!grepl("\\}$", item)) item <- paste0(item, "}")

    op <- get_val(item, "op", "string")
    a_val <- get_val(item, "a", "numeric")
    b_val <- get_val(item, "b", "numeric")

    if (is.null(a_val) || is.na(a_val)) a_val <- 0
    if (is.null(b_val) || is.na(b_val)) b_val <- 0

    res <- 0
    if (!is.null(op)) {
      if (op == "add") {
        res <- a_val + b_val
      } else if (op == "mul") {
        res <- a_val * b_val
      } else if (op == "sin") {
        res <- sin(a_val)
      } else if (op == "cos") res <- cos(a_val)
    }
    results <- c(results, res)
  }

  # Output JSON array
  cat("[")
  cat(paste(format(results, scientific = FALSE, digits = 15), collapse = ", "))
  cat("]")
} else {
  # Single mode
  op <- get_val(json_str, "op", "string")
  a <- get_val(json_str, "a", "numeric")
  b <- get_val(json_str, "b", "numeric")

  # Handle missing values safely
  if (is.null(a) || is.na(a)) a <- 0
  if (is.null(b) || is.na(b)) b <- 0

  result <- 0

  if (!is.null(op)) {
    if (op == "add") {
      result <- a + b
    } else if (op == "mul") {
      result <- a * b
    } else if (op == "sin") {
      result <- sin(a)
    } else if (op == "cos") {
      result <- cos(a)
    }
  }

  cat(format(result, scientific = FALSE, digits = 15))
}

File: public/data/chapters/chapter_01_arithmetic/logic_advanced.jl

# Logic for Julia Runtime
# Supports add, mul, sin, cos
# Reads JSON from command line argument (simple parsing to avoid deps if possible, or use standard pkg)
# Julia has "JSON" package but it might not be installed by default.
# To be safe and dependency-free like the R script, we will use regex parsing for the simple input structure.

try
    args = ARGS
    if length(args) < 1
        println(stderr, "Error: No context provided")
        exit(1)
    end

    json_str = args[1]

    function get_val(json, key, type_converter)
        # Regex to find "key": value OR "key":value
        # Value can be number, string ("..."), or null
        # Julia regex: r"\"key\"\s*:\s*([^,}]+)"
        
        r = Regex("\"$key\"\\s*:\\s*([^,}]+)")
        m = match(r, json)
        
        if m === nothing
            return nothing
        end
        
        val_str = m.captures[1]
        
        # Clean up string
        val_str = strip(val_str)
        if startswith(val_str, "\"") && endswith(val_str, "\"")
            val_str = val_str[2:end-1]
        end
        
        if val_str == "null"
            return nothing
        end
        
        return type_converter(val_str)
    end

    # Helper to check for batch mode
    is_batch = occursin("\"batch\": true", json_str) || occursin("\"batch\":true", json_str)

    if is_batch
        # Batch Mode
        # Extract examples array content
        # "examples": [ ... ]
        m_ex = match(r"\"examples\"\s*:\s*\[(.*)\]"s, json_str)
        if m_ex === nothing
             # Fallback check if it spans multiple lines or complex format - simplify assumption: it's the valid part
             # Try simpler extraction if multiline regex failed (Julia regex dot matches newline with 's' flag)
             println("[]")
             exit(0)
        end
        
        content = m_ex.captures[1]
        
        # Split objects by "},{" 
        # This is a bit fragile but works for controlled inputs without nested objects
        items = split(content, r"\}\s*,\s*\{")
        
        results = Float64[]
        
        for item in items
            # Re-add braces for regex context if missing
            full_item = item
            if !startswith(strip(full_item), "{") full_item = "{" * full_item end
            if !endswith(strip(full_item), "}") full_item = full_item * "}" end
            
            op_local = get_val(full_item, "op", x -> String(x))
            a_local = get_val(full_item, "a", x -> parse(Float64, x))
            b_local = get_val(full_item, "b", x -> parse(Float64, x))
            
            a_v = a_local === nothing ? 0.0 : a_local
            b_v = b_local === nothing ? 0.0 : b_local
            
            res = 0.0
            if op_local == "add"
                res = a_v + b_v
            elseif op_local == "mul"
                res = a_v * b_v
            elseif op_local == "sin"
                res = sin(a_v)
            elseif op_local == "cos"
                res = cos(a_v)
            end
            push!(results, res)
        end
        
        # Format output as JSON array
        # JSON.jl not assumed, manual formatting
        print("[")
        for (i, r) in enumerate(results)
            if i > 1 print(", ") end
            if isinteger(r)
                print(Int(r))
            else
                print(r)
            end
        end
        println("]")
        
    else
        # Single Mode
        op = get_val(json_str, "op", x -> String(x))
        a_val = get_val(json_str, "a", x -> parse(Float64, x))
        b_val = get_val(json_str, "b", x -> parse(Float64, x))

        # Defaults
        a = a_val === nothing ? 0.0 : a_val
        b = b_val === nothing ? 0.0 : b_val

        result = 0.0

        if op == "add"
            result = a + b
        elseif op == "mul"
            result = a * b
        elseif op == "sin"
            result = sin(a)
        elseif op == "cos"
            result = cos(a)
        else
            # Default to 0 if unknown (safe fallback for tests)
            result = 0.0
        end

        # Print result
        if isinteger(result)
            println(Int(result))
        else
            println(result)
        end
    end

catch e
    # println(stderr, "Error in Julia execution: $e")
    # Output something that won't crash the JSON parser upstream, or exit 1
    exit(1)
end
